File: code/trees/bst-traversal.py
# The three recursive traversals are ONE recursion; only the line where you
# visit(t) moves. In-order of a BST visits the keys in SORTED order.
def inorder(t):
    if t is None:
        return
    # visit(t)            # PRE-order:  node BEFORE its subtrees
    inorder(t.left)
    visit(t)              # IN-order:   node BETWEEN the subtrees (sorted!)
    inorder(t.right)
    # visit(t)            # POST-order: node AFTER its subtrees

File: code/trees/bst-traversal.rs
// The three recursive traversals are ONE recursion; only the line where you
// visit(n) moves. In-order of a BST visits the keys in SORTED order.
fn inorder(t: &Option<Box<Node>>) {
    if let Some(n) = t {
        // visit(n);        // PRE-order:  node BEFORE its subtrees
        inorder(&n.left);
        visit(n);           // IN-order:   node BETWEEN the subtrees (sorted!)
        inorder(&n.right);
        // visit(n);        // POST-order: node AFTER its subtrees
    }
}
